sqlmesh/core/snapshot/evaluator.py

```diff
@@ -543,14 +544,20 @@ def apply(query_or_df: QueryOrDF, index: int = 0) -> None:
 
             if limit is not None:
                 query_or_df = next(queries_or_dfs)
-                if isinstance(query_or_df, exp.Select):
-                    existing_limit = query_or_df.args.get("limit")
-                    if existing_limit:
-                        limit = min(
-                            limit,
-                            execute(exp.select(existing_limit.expression)).rows[0][0],
-                        )
-                return query_or_df.head(limit) if hasattr(query_or_df, "head") else self.adapter._fetch_native_df(query_or_df.limit(limit))  # type: ignore
+                if isinstance(query_or_df, PyDataFrame):
+                    return query_or_df.limit(limit)
+                if isinstance(query_or_df, pd.DataFrame):
+                    return query_or_df.head(limit)
+
+                assert isinstance(query_or_df, exp.Query)
```

Comment:
I think there was a bug here previously, which was silenced due to `type: ignore`. When `query_or_df` is a Spark dataframe, `head` will [return](https://spark.apache.org/docs/3.1.1/api/python/reference/api/pyspark.sql.DataFrame.head.html#pyspark.sql.DataFrame.head) a `Row` or a list of `Row`s, i.e. _not_ a `DF`, and so invariants may break. E.g. in https://github.com/TobikoData/sqlmesh/blob/9e19b98338ca97e6aefd1525256d325371bfee9f/sqlmesh/core/snapshot/evaluator.py#L173 we'd return a different object from what the method's type hint says.

Reply:
The to ensure we have a `Query` happened because the type union also contains `DerivedTable` for some reason (maybe to take CTEs and / or UDTFs into account? not sure..). Only `Query` expressions support the `limit` builder, so if for whatever reason `query_or_df` was not a `Query`, the `.limit(...)` would crash.